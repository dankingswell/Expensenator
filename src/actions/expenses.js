import uuid from "uuid"
//Actions
// Add, remove,expense

//Uses defualt object for params

export const addExpense = ({description ="",
                            note = "" ,
                            amount=0,
                            createdAt=0  
                                } ={}) => 
    ({
    type:"ADD_EXPENSE",
    expenses: {
        id:uuid(),
        description,
        note,
        createdAt,
        amount
    }
})

export const editExpense = ( id, updates) => (
    {
        type:"EDIT_EXPENSE",
        id,
        updates
    }
)

export const removeExpense = ( {id=0} = {}

) =>({
    type:"REMOVE_EXPENSE",
    id
})
