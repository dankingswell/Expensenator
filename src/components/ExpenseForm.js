import React from "react";
import moment from "moment";
import {connect} from "react-redux"
import "react-dates/initialize"

moment().locale('en-gb');
import {SingleDatePicker} from "react-dates";

class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        
        this.state = {
            description:props.expense ? props.expense.description: "",
            note: props.expense ? props.expense.note:"",
            amount:props.expense ? (props.expense.amount / 100).toString() :"",
            createdAt:props.expense ? moment(props.expense.createdAt) :moment(),
            Calendarfocused:false,
            error:false
        }
        
    }
    
    onDescriptionChange = (event) => {
        const description =  event.target.value;
        this.setState(()=>({description}))
    }
    onNoteChange = (event) => {
        const note = event.target.value;
        this.setState( ()=>({note} ))
    }
    onAmountChange = (event) =>{
        const amount = event.target.value;
        if(amount.match(/^\d*(\.\d{0,2})?$/)){
        this.setState(()=>({amount}));
        }else{
            event.target.value = Math.round(amount*100)/100;
        }
    }
    onDateChange = (createdAt) => {
        if(createdAt){
        this.setState(()=>({createdAt}))
        }
    }

    onFocusChange = ({focused}) =>{
        this.setState(()=>({Calendarfocused:focused}))
    }

    onSubmit= (event) => {
        event.preventDefault()
        
        if(!this.state.description || !this.state.amount)
        {
            this.setState(()=>({error:true}))
        }else{
            this.setState(()=>({error:false}))
            this.props.onSubmit({
                description: this.state.description,
                amount:parseFloat(this.state.amount,10)*100,
                createdAt: this.state.createdAt.valueOf(),
                note:this.state.note
            })
        }
        
    }
    render(){
        
        return (
        <div>
            <form
            onSubmit={this.onSubmit}
            >
                <input
                type="text"
                placeholder="Description"
                autoFocus
                value ={this.state.description}
                onChange={this.onDescriptionChange}
                />

                <input
                type="text"
                value={this.state.amount}
                placeholder={"Amount"}
                onChange={this.onAmountChange}
                />

                <SingleDatePicker 
                date ={this.state.createdAt}
                onDateChange={this.onDateChange}
                focused={this.state.Calendarfocused}
                onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={(day)=>(false)}
                />
                
                <textarea
                placeholder={"Please enter any notes here"}
                value={this.state.note}
                onChange={this.onNoteChange}
                ></textarea>

                <button>Add Expense</button>
                {this.state.error && <p>Please complete the description and amount fields</p>}
                
                
            </form>
        </div>
    )}

}

export default ExpenseForm;