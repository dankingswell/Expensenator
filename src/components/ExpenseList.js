import React from "react"
import { connect } from "react-redux"
import ExpenseListItem from "./ExpenseListItem"
import selectExpenses from "../selectors/expenses"

// export as named to allow for testing
export const ExpenseList = (props) => (
    
    <div>
    
        <h1>ExpenseList</h1>
        {
            props.expenses.length > 0 ? 
             <div>   
                {props.expenses.map((expense)=>{
                    
                    return <ExpenseListItem 
                    key={expense.id}
                    // passes all props as KVP's
                    {...expense}
                    />
                })}
            </div> : <p>All cleared, no expenses</p>
        }
    </div>
);


// function is passed to connect to communicate
// which parts of state need to be given to component
// connect implicitly passes state
const mapStateToProps =(state) => {
    return {
        expenses:selectExpenses(state.expenses,state.filters)
    }
}


// connect returns a wrapper function
// this wrapper function has the ability to use the state passed in
// invoke new function with state on component
// component is given state props
export default connect(mapStateToProps)(ExpenseList);

// export connected component