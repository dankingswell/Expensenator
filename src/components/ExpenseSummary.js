import React from "react";
import {connect} from "react-redux";
import getTotalOfCurrentExpenses from "../selectors/expenseHeader"
import selectExpenses from "../selectors/expenses"

export const ExpenseHeader = (props) => (
    <div>
        <p>
            {`Showing ${props.numberOfExpenses}
             ${props.numberOfExpenses > 1 ? "expenses" : "expense"}
             , Total amount: Â£${props.total}`} 
       </p>
    </div>
);

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses,state.filters)
    return{
        total:getTotalOfCurrentExpenses(visibleExpenses),
        numberOfExpenses:visibleExpenses.length
    }
}

export default connect(mapStateToProps)(ExpenseHeader);