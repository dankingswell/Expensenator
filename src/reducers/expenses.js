// Expenses reducer
//takes default state and action

const expenseReducerDefaultState = []
const expensesReducer = (state = expenseReducerDefaultState,  action) =>{

    switch(action.type){
        case "ADD_EXPENSE":
        // ES6 SpreadOperator =  state.concat(action.expenses)
        // [...Arrayname, "new Prop"]  ... means take all vals from array mentioned      
        return [...state, action.expenses];
        
        case "REMOVE_EXPENSE":
            return state.filter((item)=>{
                return item.id != action.id   
            });
        
        case "EDIT_EXPENSE":
            return state.map((expense)=>{
                if (expense.id == action.id ){
                    return {
                        ...expense,
                        // as the functions passes an object it acts as a KVP
                        ...action.updates
                       }
                    }else{
                    return expense;
                }
            });

        default:
            return state;
    }

}

export default expensesReducer;