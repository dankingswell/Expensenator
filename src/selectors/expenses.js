import moment from "moment"

const getVisibileExpenses = (expenses,{text, sortBy ,startDate,endDate}) =>{
    // filter expense and if all fields match then return value
    return expenses.filter( (expense) => {
       
        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day"):true;
        const endDateMatch = endDate  ?  endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch = typeof(text) != "string" ||  expense.description.toLowerCase().includes(text.toLowerCase()) ;
        return startDateMatch && endDateMatch && textMatch;
    } )
    // sorting return data
    .sort( (a, b) => {
        if(sortBy == "date"){
            return a.createdAt < b.createdAt ? 1 : -1
        }
        if(sortBy == "amount"){
            return a.amount > b.amount ? -1 : 1;
        }
    });
}

export default getVisibileExpenses; 