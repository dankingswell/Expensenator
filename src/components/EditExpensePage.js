import React from "react"
import {connect} from "react-redux";
import ExpenseForm from "./ExpenseForm"
import {editExpense, removeExpense} from "../actions/expenses"

// routes are passed props implicitly but can be accessed via props
const editExpensePage = (props) => (
    
    <div>
        <h1>Edit Expense</h1>
        <ExpenseForm 
        expense={props.expense}
        onSubmit={(expenseReturn)=>{
            props.dispatch(editExpense(props.expense.id, expenseReturn ))
            props.history.push("/")
        }}
  />
        <button 
        onClick={ () =>{
            props.dispatch(removeExpense({id:props.expense.id}));
            props.history.push("/");
            }
        }>Remove Expense
        </button>

    </div>

);

// history allows us to 
// redirect, go back, go foward, match 
// location: current URL values e.g. ? query="whattt" will be inside location 
const mapStateToProps = (state,props) =>{
    return {
        expense:state.expenses.find(
            (Expense)=>{
                return Expense.id == props.match.params.id
            }
        )
    }
}
export default connect(mapStateToProps)(editExpensePage);