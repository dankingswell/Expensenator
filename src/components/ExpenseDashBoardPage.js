import React from "react"
import ExpenseList from "./ExpenseList"
import ExpenseListFilters from "./ExpenseListFilters"
import ExpenseHeader from "./ExpenseHeader"




const expenseDashBoardPage = () => (
    <div>
        <ExpenseHeader />
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);


export default expenseDashBoardPage;