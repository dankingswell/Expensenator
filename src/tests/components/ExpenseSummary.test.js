import React from "react"
import {ExpenseHeader} from "../../components/ExpenseSummary"
import {shallow} from "enzyme"
import expenses from "../fixture/expenses"

test("summary header should render correctly", () => {
	const wrapper = shallow(<ExpenseHeader expenses={[]}/>);
	expect(wrapper).toMatchSnapshot();
});