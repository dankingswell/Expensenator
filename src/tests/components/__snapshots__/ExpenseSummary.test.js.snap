// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`summary header should render correctly 1`] = `
<div>
  <p>
    Showing undefined
             expense
             , Total amount: Â£undefined
  </p>
</div>
`;
